python/cuspatial/cuspatial/geometry/geoarrowbuffers.py

```diff
@@ -0,0 +1,558 @@
+        )
+
+
+class CoordinateArray:
+    def __init__(self, xy, z=None, data_locale=cudf):
+        """
+        A GeoArrow column of points. The CoordinateArray stores all of the
+        points within a single data source, typically a cuspatial.GeoSeries,
+        in the format specified by GeoArrow.
+        """
+        self.data_location = data_locale
+        self.xy = xy
+        if z is not None:
+            self.z = z
+
+    @property
+    def data_location(self):
+        """
+        A `Series` supporting python library, expected to be either `cudf` or
+        `pandas`. If `cudf`, GeoArrowBuffer is stored on the GPU and used in
+        cuspatial operations. If `pandas`, GeoArrowBuffer is stored on Host and
+        used for creating GeoPandas objects.
+        """
+        return self._data_location
+
+    @data_location.setter
+    def data_location(self, data_location):
+        if data_location not in [cudf, pd]:
+            raise NotImplementedError(
+                "only cudf and pandas CoordinateArrays "
+                "are supported at this time"
+            )
+        else:
+            self._data_location = data_location
+
+    def _serialize(self, data):
+        try:
+            if hasattr(data, "to_pandas"):
+                return data.to_pandas()
+            else:
+                return cudf.Series(data)
+        except Exception as e:
+            raise TypeError(e, f"{type(data)} is not supported in cuspatial")
+
+    @property
+    def xy(self):
+        """
+        The coordinates of this Geometry Column in interleaved format
+        [x,y,...,x,y].
+        """
+        return self._xy
+
+    @xy.setter
+    def xy(self, xy):
+        if (len(xy) % 2) != 0:
+            raise ValueError("xy must have even length")
+        self._xy = self._serialize(xy)
+
+    @property
+    def z(self):
+        """
+        An optional third dimension for this Geometry.
+        """
+        if hasattr(self, "_z"):
+            return self._z
+        else:
+            return None
+
+    @z.setter
+    def z(self, z):
+        self._z = self._serialize(z)
+
+    def __getitem__(self, index):
+        if isinstance(index, slice):
+            new_index = slice(index.start * 2, index.stop * 2 + 1, index.step)
+            return self.xy.iloc[new_index]
+        return self.xy.iloc[(index * 2) : (index * 2) + 2]
+
+    def __repr__(self):
+        return f"xy:\n" f"{self.xy.__repr__()}\n"
+
+    def copy(self, deep=True):
+        """
+        Create a copy of all points.
+
+        Parameters
+        ----------
+        deep : Boolean.
+            If set to False, a new object referencing the same
+            set of GPU memory will be created. If True, GPU memory will be
+            copied.
+        """
+        if hasattr(self, "_z"):
+            z = self.z.copy(deep)
+        else:
+            z = None
+        result = type(self)(self.xy.copy(deep), z)
+        return result
+
+    @property
```

Comment:
Any benefit to making these `cached_properties`? Or is the additional memory overhead too much?

Reply:
Do we support cached property on GPU? This is a `cudf.Series`

Reply:
Yes -- it shouldn't matter whether the returned object lives on the CPU or GPU. Here's how `cached_property` works; consider the cached property `bar` below:

```python
class Foo:
    @cached_property
    def bar(self):
        return cudf.Series([1, 2, 3])

>>> f = Foo()
>>> f.bar  # invokes bar() method & makes `bar` a regular attribute of the class
cudf.Series([1, 2, 3])
>>> f.bar  # no method invoked, `bar` is just a regular attribute
cudf.Series([1,  2, 3])
```

So, the first time a method decorated with `cached_property` is accessed, the body of the method is executed. After that, the method disappears, and it is replaced by an attribute of the same name, and containing the result of the first invocation.

Reply:
cool, done

Reply:
This is still a `@property` as opposed to `@cached_property` for me still. Am I missing something?